### PR TITLE
Package shared tce-core dependencies

### PR DESCRIPTION
**Pull request by MiroDojkic**
### This PR
- moves dependencies that are shared between tce-core components or tce-core module and Tailor to distributable packages
- this decoupling aims to enable complete separation of tce-core from Tailor

Note: [Lerna](https://github.com/lerna/lerna) is used for convenience of building, symlinking etc. of multiple packages. Dev dependencies are not hoisted so each package can be easily moved to a separate repo if we decide so.

### CODE REVIEW

File: server/repository/repository.model.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
 'use strict';
 
-const { getRepositoryRelationships, getSchema } = require('../../config/shared/activities');
 const { Model } = require('sequelize');
 const pick = require('lodash/pick');
 const Promise = require('bluebird');
+const { schema } = require('@tailor/config');
 
+const { getRepositoryRelationships, getSchema } = schema;
```
Comment (underscope) [inline server/repository/repository.model.js:8]: Consider adding one blank line below (before class definition) :bowtie:

File: packages/vue-radio/package.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    "type": "git",
+    "url": "git+https://github.com/ExtensionEngine/tailor.git"
+  },
+  "scripts": {
+    "prebuild": "del-cli dist",
+    "build": "bili",
+    "lint": "eslint --ext .js,.vue ."
+  },
+  "bugs": {
+    "url": "https://github.com/ExtensionEngine/vue-radio/issues"
```
Comment (underscope) [inline packages/vue-radio/package.json]: *

File: packages/vue-radio/package.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  "version": "0.0.0",
+  "description": "Vue radio plugin",
+  "author": "ExtensionEngine <info@extensionengine.com>",
+  "homepage": "",
+  "license": "MIT",
+  "main": "dist/vue-radio.js",
+  "module": "dist/vue-radio.esm.js",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/ExtensionEngine/tailor.git"
```
Comment (underscope) [inline packages/vue-radio/package.json:11]: *

File: packages/vue-radio/package.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "@extensionengine/vue-radio",
+  "version": "0.0.0",
```
Comment (underscope) [inline packages/vue-radio/package.json]: 0.0.1 ?

File: packages/config/package.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@tailor/config",
+  "version": "0.0.0",
```
Comment (underscope) [inline packages/config/package.json]: 0.0.1

File: client/components/content-containers/tcc-assessment-pool/edit/index.vue [hunk 6]
```diff
@@ -34,11 +34,10 @@
 </template>
 
 <script>
-import AddElement from 'tce-core/AddElement';
```
Comment (underscope) [inline client/components/content-containers/tcc-assessment-pool/edit/index.vue:37]: Assessment poo and exam should be omitted and installed as separate package (extracted)

File: client/components/catalog/Card/index.vue [hunk 7]
```diff
@@ -97,14 +96,15 @@ const getPublishingInfo = hasChanges => hasChanges
...
+  inject: ['$schemaService'],
   props: {
     repository: { type: Object, required: true }
   },
   data: () => ({ isSchemaTruncated: false }),
   computed: {
     name: ({ repository }) => repository.name,
     description: ({ repository }) => repository.description,
-    schema: ({ repository }) => getSchema(repository.schema).name,
+    schema: ({ $schemaService, repository }) => $schemaService.getSchema(repository.schema).name,
```
Comment (underscope) [inline client/components/catalog/Card/index.vue]: Watch the line length limitations ;)
` schema: vm => vm.$schemaService.getSchema(vm.repository.schema).name`

File: client/components/editor/Sidebar/Navigation.vue [hunk 8]
```diff
@@ -64,12 +66,14 @@ export default {
...
-    activityConfigs: vm => getOutlineLevels(vm.repository.schema),
+    activityConfigs: vm => vm.$schemaService.getOutlineLevels(vm.repository.schema),
     activityTypes: vm => vm.activityConfigs.map(it => it.type),
     editableTypes: vm => vm.editableActivityConfigs.map(it => it.type),
     editableActivityConfigs() {
-      return this.activityConfigs.filter(it => isEditable(it.type));
+      return this.activityConfigs.filter(it => this.$schemaService.isEditable(it.type));
+    },
+    activityTree() {
+      return toTreeFormat(this.activities, this.$schemaService, this.editableActivityConfigs);
```
Comment (underscope) [inline client/components/editor/Sidebar/Navigation.vue]: It seems like the third argument no longer makes sense (since it can be all done within`toTreeFormat`)

File: client/components/repository/common/CopyActivity/RepositoryTree.vue [hunk 9]
```diff
@@ -36,18 +36,21 @@
...
+  inject: ['$schemaService'],
   props: {
     schemaName: { type: String, required: true },
     activities: { type: Array, required: true },
     supportedLevels: { type: Array, required: true }
   },
   data: () => ({ selected: [], search: '' }),
   computed: {
-    activityTree: vm => toTreeFormat(vm.activities, vm.supportedLevels),
+    activityTree: vm => toTreeFormat(vm.activities, vm.$schemaService, vm.supportedLevels),
```
Comment (underscope) [inline client/components/repository/common/CopyActivity/RepositoryTree.vue]: A bit lengthy

File: client/components/repository/common/Sidebar/Relationship.vue [hunk 10]
```diff
@@ -79,7 +79,7 @@ export default {
     groupedOptions() {
       const grouped = groupBy(this.options, 'type');
       return flatMap(grouped, (it, type) => {
-        const headerLabel = pluralize(getLevel(type).label);
+        const headerLabel = pluralize(this.$schemaService.getLevel(type).label);
```
Comment (underscope) [inline client/components/repository/common/Sidebar/Relationship.vue]: Consider doing pluralization within concat line (to make this one a bit simpler)

File: packages/components/src/components/QuestionContainer/Controls.vue [hunk 11]
```diff
@@ -23,6 +23,7 @@
 
 <script>
 export default {
+  name: 'controls',
```
Comment (underscope) [inline packages/components/src/components/QuestionContainer/Controls.vue]: consider renaming it to question-controls

File: packages/utils/src/activity.js [hunk 12]
```diff
@@ -44,12 +32,19 @@ export function getAncestors(activities, activity) {
   return [...ancestors, parent];
 }
 
-export function toTreeFormat(activities, targetLevels, parentId = null, level = 1) {
-  return getOutlineChildren(activities, parentId).map(activity => ({
+export function getOutlineChildren(activities, schema, parentId) {
```
Comment (underscope) [inline packages/utils/src/activity.js]: Since there is an outline activity getter it might make sense to keep the previous version that is relying on getting only outlineActivities as first param (should be renamed)

File: packages/utils/src/activity.js [hunk 13]
```diff
@@ -44,12 +32,19 @@ export function getAncestors(activities, activity) {
...
+  const children = getChildren(activities, parentId);
+  if (!parentId || !children.length) return children;
+  const types = schema.getLevel(find(activities, { id: parentId }).type).subLevels;
+  return filter(children, it => types.includes(it.type));
+}
+
+export function toTreeFormat(activities, schema, targetLevels, parentId = null, level = 1) {
```
Comment (underscope) [inline packages/utils/src/activity.js]: The new signature now feels bloated, which relates to the comment above. It is also additionally limiting now since one might want to do this for part of the subtree (i.e. list containers) 🤔

File: packages/config/src/index.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+/* eslint-disable */
+const defaultConfiguration = require('./activities-rc');
+const customConfiguration = require('./activities-rc.load')();
+/* eslint-enable */
+if (!process.env.ENABLE_DEFAULT_SCHEMA && !isEmpty(customConfiguration)) {
+  defaultConfiguration.SCHEMAS = [];
+}
+
+const { SCHEMAS } = mergeConfig(defaultConfiguration, customConfiguration);
```
Comment (underscope) [inline packages/config/src/index.js:15]: Overall feels a bit convoluted. We should probably make it a bit more readable (separate PR)

File: package.json [hunk 15]
```diff
@@ -12,6 +12,10 @@
   "scripts": {
     "dev:client": "poi --serve",
     "build": "del-cli node_modules/.cache/vue-loader && poi --prod",
+    "tce:bootstrap": "lerna bootstrap",
+    "tce:build": "lerna run build",
+    "tce:link": "npm link ./packages/*",
+    "tce:setup": "npm run tce:bootstrap && npm run tce:build && npm run tce:link",
```
Comment (underscope) [inline package.json]: Should be documented within readme.
Should we facilitate installing all dependencies using a single command to simplify running?

File: packages/components/package.json [hunk 16]
```diff
@@ -0,0 +1,62 @@
+{
+  "name": "@tailor/components",
```
Comment (underscope) [inline packages/components/package.json]: Consider renaming to core-components (since content elements are as well components)

File: packages/utils/src/activity.js [hunk 17]
```diff
@@ -32,19 +32,32 @@ export function getAncestors(activities, activity) {
...
-  const types = schema.getLevel(find(activities, { id: parentId }).type).subLevels;
+  const parentType = find(activities, { id: parentId }).type;
+  const types = schema.getLevel(parentType).subLevels;
   return filter(children, it => types.includes(it.type));
 }
 
-export function toTreeFormat(activities, schema, parentId = null, level = 1) {
-  return getOutlineChildren(activities, schema, parentId).map(activity => ({
+export function getOutlineTree(activities, schema, options = {}) {
+  return toTreeFormat(activities, schema, { filterFn: getOutlineChildren, ...options });
```
Comment (ikovac) [inline packages/utils/src/activity.js]: @underscope is this a good solution? 
I'm a little bit skeptical about passing down the filter function to the recursive function because the consumer will have obligation to pass the correct filter function that doesn't make infinite recursive calls. What do you think?

Comment (underscope) [reply packages/utils/src/activity.js]: We can have a depth parameter and throw an error if it's exceeded? 😉

File: packages/utils/src/activity.js [hunk 18]
```diff
@@ -44,12 +32,30 @@ export function getAncestors(activities, activity) {
...
+export function toTreeFormat(activities, { filterNodesFn, processNodeFn }, _internals = {}) {
+  const { parentId = null, level = 1, maxLevel = 20 } = _internals;
+  if (level > maxLevel) throw new Error('Max level exceeded');
+  return filterNodesFn(activities, parentId).map(activity => ({
     ...activity,
     name: activity.data.name,
     level,
-    selectable: !!targetLevels.find(it => it.type === activity.type),
-    children: toTreeFormat(activities, targetLevels, activity.id, level + 1)
+    selectable: processNodeFn ? processNodeFn(activity) : false,
```
Comment (underscope) [inline packages/utils/src/activity.js]: Consider not having this baked in, make processNodeFn return object that will be spread (should be last)

File: packages/utils/src/activity.js [hunk 19]
```diff
@@ -44,12 +32,30 @@ export function getAncestors(activities, activity) {
...
-  return getOutlineChildren(activities, parentId).map(activity => ({
+export function getOutlineChildren(activities, parentId, schema) {
+  const children = getChildren(activities, parentId);
+  if (!parentId || !children.length) return children;
+  const parentType = find(activities, { id: parentId }).type;
+  const types = schema.getLevel(parentType).subLevels;
+  return filter(children, it => types.includes(it.type));
+}
+
+export function getOutlineChildrenFilterFn(schema) {
```
Comment (underscope) [inline packages/utils/src/activity.js]: Consider not using the same function for resolving children in the hierarchy and filtering out specific types. That way you don't have to pass the filter function always and you don't have to have a more specific signature expecting parentId. To recap, instead of:
`return filterNodesFn(activities, parentId)`
you can do
`const parentActivities = filter(activities, { parentId })`
and then apply filtering. Instead of getOutlineChildrenFilterFn
you can pass isOutlineAcitivty which is already exposed within the schema config package.

File: packages/utils/src/activity.js [hunk 20]
```diff
@@ -44,12 +32,30 @@ export function getAncestors(activities, activity) {
   return [...ancestors, parent];
 }
 
-export function toTreeFormat(activities, targetLevels, parentId = null, level = 1) {
-  return getOutlineChildren(activities, parentId).map(activity => ({
+export function getOutlineChildren(activities, parentId, schema) {
```
Comment (underscope) [inline packages/utils/src/activity.js]: schema config module might be a better place (signature might be simpler, no need to pass schema)

**Comment by underscope** [comment]
@ikovac Don't forget to merge latest changes on develop ;)

File: client/components/editor/Sidebar/Navigation.vue [hunk 21]
```diff
@@ -64,12 +66,22 @@ export default {
...
   },
   computed: {
-    activityTree: vm => toTreeFormat(vm.activities, vm.editableActivityConfigs),
-    activityConfigs: vm => getOutlineLevels(vm.repository.schema),
+    activityConfigs: vm => vm.$schemaService.getOutlineLevels(vm.repository.schema),
     activityTypes: vm => vm.activityConfigs.map(it => it.type),
     editableTypes: vm => vm.editableActivityConfigs.map(it => it.type),
     editableActivityConfigs() {
-      return this.activityConfigs.filter(it => isEditable(it.type));
+      return this.activityConfigs.filter(it => this.$schemaService.isEditable(it.type));
```
Comment (underscope) [inline client/components/editor/Sidebar/Navigation.vue]: Some of these seem no longer needed 🤔 ? (editableActivityConfigs, editableTypes, activityTypes, activityConfigs)

File: packages/config/src/schema.js [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+const reduce = require('lodash/reduce');
+const sortBy = require('lodash/sortBy');
+const union = require('lodash/union');
+const uniq = require('lodash/uniq');
+
+module.exports = (schemas, defaultConfiguration) => {
+  return {
+    getSchemaId,
+    getSchema,
+    getLevel: getActivityConfig,
```
Comment (underscope) [inline packages/config/src/schema.js:20]: Please introduce getLevel deprecation and use getActivityConfig (Keep alias and display warning, separate PR)

File: packages/config/src/schema.js [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+    getOutlineLevels,
```
Comment (underscope) [inline packages/config/src/schema.js:21]: Please rename to getOutlineConfig and deprecate (Keep alias and display warning, separate PR)

File: packages/config/src/schema.js [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+    isOutlineActivity,
+    getOutlineChildren,
+    filterOutlineActivities,
```
Comment (underscope) [inline packages/config/src/schema.js:24]: There could be one potential issue with this one, let's chat.

File: packages/config/src/schema.js [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    isTrackedInWorkflow,
+    getRepositoryMetadata,
+    getActivityLabel,
+    getActivityMetadata,
+    getElementMetadata,
+    getLevelRelationships,
```
Comment (underscope) [inline packages/config/src/schema.js:30]: Should be renamed to getActivityRelationships (Separate PR)

File: packages/config/src/schema.js [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+    getRepositoryRelationships,
+    getSiblingTypes,
+    getSupportedContainers,
+    getContainerTemplateId,
+    hasAssessments: level => getActivityConfig(level).hasAssessments,
```
Comment (underscope) [inline packages/config/src/schema.js]: Should be no longer used. Please check.

File: packages/config/src/schema.js [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+    isEditable: activityType => {
+      const config = getActivityConfig(activityType);
+      const hasContainers = !!getSupportedContainers(activityType).length;
+      return hasContainers || config.hasAssessments;
+    }
+  };
+
+  function getSchemaId(type) {
```
Comment (underscope) [inline packages/config/src/schema.js:42]: Please document all exposed config utils (JSDoc; Can be a separate PR)

File: client/components/editor/Sidebar/Navigation.vue (hunk 21 above)
Comment (ikovac) [reply client/components/editor/Sidebar/Navigation.vue]: Yea, `activityTypes` is not used anymore

Comment (underscope) [reply client/components/editor/Sidebar/Navigation.vue]: Since schema service has isEditable method, isActivityEditable is also not needed in this form, which also makes editableTypes and editableActivityConfigs obsolete ? 🤔

Comment (ikovac) [reply client/components/editor/Sidebar/Navigation.vue]: You are right, I only checked usage instead of the actual need for a particular code 🤦‍♂️.

**Comment by kjuej** [comment]
@ikovac @underscope Issue with copying assessment element:
Screen Shot 2021-05-11 at 16 48 42
Checked on develop and seems to be functioning correctly there.

**Comment by ikovac** [comment]
@kjuej I can't reproduce it. Is there any console error?

**Comment by kjuej** [comment]
> @kjuej I can't reproduce it. Is there any console error?

No console error, might be an issue on built version?

**Comment by ikovac** [comment]
It works for me both locally with `npm run packages:setup` and via `npm i`. We can sync and try to debug it together whenever you are free.

**Comment by ikovac** [comment]
@kjuej I spotted where the bug is, can you please test it again?  ty, good catch! 🚀
